src/inmanta/agent/agent.py

```diff
@@ -1526,3 +1527,133 @@ async def get_facts(self, env: uuid.UUID, agent: str, resource: dict[str, Any])
+        self.path = storage
+        self.thread_pool = threadpool
+
+    async def create_and_install_environment(self, blueprint: EnvBlueprint) -> None:
+        loop = asyncio.get_running_loop()
+        req = list(blueprint.requirements)
+        self.env.init_env()
+        install_for_config = functools.partial(
+            self.env.install_for_config,
+            requirements=list(pkg_resources.parse_requirements(req)),
+            config=blueprint.pip_config,
+            should_be_active=False,
+        )
+        await loop.run_in_executor(self.thread_pool, install_for_config)
+
+
+class VirtualEnvironmentManager:
+    def __init__(self):
+        self._environment_map: dict[EnvBlueprint, ExecutorVirtualEnvironment] = {}
+
+    def create_env_storage(self, storage: str) -> str:
+        envs_dir = storage
+        new_env_dir = os.path.join(envs_dir, str(uuid.uuid4()))
+        os.mkdir(new_env_dir)
+        return new_env_dir
+
+    async def create_environment(
+        self, blueprint: EnvBlueprint, storage: str, threadpool: ThreadPoolExecutor
+    ) -> ExecutorVirtualEnvironment:
+        # create a new storage location for the new environment
+        env_storage = self.create_env_storage(storage)
+        process_environment = ExecutorVirtualEnvironment(env_storage, threadpool)
+        await process_environment.create_and_install_environment(blueprint)
+        self._environment_map[blueprint] = process_environment
+        return process_environment
+
+    async def get_environment(
+        self, blueprint: EnvBlueprint, storage: str, threadpool: ThreadPoolExecutor
+    ) -> ExecutorVirtualEnvironment:
+        if blueprint in self._environment_map:
+            return self._environment_map[blueprint]
+        return await self.create_environment(blueprint, storage, threadpool)
+
+
+@dataclasses.dataclass(frozen=True)
+class ExecutorId:
+    agent_name: str
+    config_model_version: int
+
+
+class Executor:
+    def __init__(self, executor_id: ExecutorId, executor_virtual_env: ExecutorVirtualEnvironment, storage: str):
+        self.executor_id = executor_id
+        self.executor_virtual_env: ExecutorVirtualEnvironment = executor_virtual_env
+        self.storage = storage
+
+    def load_code(self, sources: Sequence["ModuleSource"], threadpool: ThreadPoolExecutor):
+        print("Load the code of sources for executor")
+
+    def execute(self, resources: list[Resource]):
+        print("Start deploy of resources")
+
+
+class ExecutorManager:
+    def __init__(self, agent: Agent, environment_manager: VirtualEnvironmentManager):
+        self.executor_map: dict[ExecutorId, Executor] = {}
+        self.environment_manager = environment_manager
+        self.agent = agent  # for now the executorManager lives in an agent
+        self.storage = self.create_storage()
+
+    async def create_executor(self, agent_name: str, config_model_version: int, blueprint: EnvBlueprint) -> Executor:
```

Comment:
I don't think we want to restart the executor for every version, only if the actual code changes.